### PR TITLE
ipc4: set pipeline components dir based on source component

### PR DESCRIPTION
**Pull request by kfrydryx**
Prefer to use source components direction, rather than dai direction when setting pipeline components direction. This is required for host copier to host copier pipelines to be handled correctly. `pipeline_get_dai_comp` assumed last component in a pipeline to be DAI and returns it even if it is not a DAI, so handling of such pipelines was improper. 

Signed-off-by: Krzysztof Frydryk <krzysztofx.frydryk@intel.com>

### CODE REVIEW

**Comment by keqiaozhang** [comment]
SOFCI TEST

File: src/ipc/ipc4/handler.c [hunk 1]
```diff
@@ -393,18 +393,31 @@ static int ipc_wait_for_compound_msg(void)
...
+			dir_src = pipe->pipeline->source_comp;
+			break;
+		} else if (pipe->pipeline->sink_comp->direction_set) {
+			dir_src = pipe->pipeline->sink_comp;
+			break;
+		}
+	}
+	if (!dir_src) {
+		tr_info(&ipc_tr, "no direction source in pipeline");
 		return 0;
```
Comment (RanderWang) [inline src/ipc/ipc4/handler.c:421]: direction is vital. Could you help to refine it with a error msg and return IPC4_INVALID_RESOURCE_STATE. Thanks!

Comment (kfrydryx) [reply src/ipc/ipc4/handler.c]: sure, changed to error.

**Review by RanderWang — APPROVED** [review]
LGTM

**Review by marcinszkudlinski — APPROVED** [review]

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**